apps/passport-server/src/routing/routes/pcdRoutes.ts

```diff
@@ -1,19 +1,50 @@
-import { ProveRequest, VerifyRequest } from "@pcd/passport-interface";
+import {
+  hashRequest,
+  PendingStampPCD,
+  ProveRequest,
+  StampPCDStatus,
+  VerifyRequest,
+} from "@pcd/passport-interface";
 import express, { NextFunction, Request, Response } from "express";
-import { getSupportedPCDTypes, prove, verify } from "../../services/proving";
-import { ApplicationContext } from "../../types";
+import {
+  getSupportedPCDTypes,
+  initPackages,
+  prove,
+  verify,
+} from "../../services/proving";
+import { ServerProvingContext } from "../../types";
 
-export function initPCDRoutes(
+export async function initPCDRoutes(
   app: express.Application,
-  _context: ApplicationContext
-): void {
+  _provingContext: ServerProvingContext
+): Promise<void> {
+  await initPackages();
+
   app.post(
     "/pcds/prove",
     async (req: Request, res: Response, next: NextFunction) => {
       try {
         const proveRequest: ProveRequest = req.body;
-        const response = await prove(proveRequest);
-        res.status(200).json(response);
+        const hash = hashRequest(proveRequest);
+
+        // don't add identical proof requests to queue to prevent accidental or
+        // malicious DoS attacks on the proving queue
+        if (!_provingContext.stampStatus.has(hash)) {
+          _provingContext.queue.push(proveRequest);
+          if (_provingContext.queue.length == 1) {
+            _provingContext.stampStatus.set(hash, StampPCDStatus.PROVING);
+            prove(proveRequest, _provingContext);
```

Comment:
this function is async, but it's not being awaited here, is that intentional? another thing is that we should probably separate the concerns a little more in this PR between the code that handles http requests and the code that manipulates the queue.

I'd create a couple of functions in the file that is responsible for the queuing called `enqueueRequest`, which returns a `PendingPCD`, and abstracts away the queue logic completely from the http code.

basically this code should not have access to `provingContext`.